### PR TITLE
[MAINTENANCE] Refactor `BaseDataContext` and `DataContext` into factory functions

### PR DESCRIPTION
**Pull request by cdkini**
Changes proposed in this pull request:
- Introduce `SerializableDataContext` to act as common ground between `FileDataContext` and `CloudDataContext`
- Update `BaseDataContext` and `DataContext` to become functions that return a subclass of `AbstractDataContext`
- Update all static/class calls to `BaseDataContext` and `DataContext`
- Update all type hints

### Definition of Done
- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/data_context/data_context/abstract_data_context.py
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py:166]: Need to move over to the base class as `BaseDataContext` can no longer inherit

File: great_expectations/data_context/data_context/abstract_data_context.py
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py:258]: Ported over from `BaseDataContext` - required for proper validation operator creation

File: great_expectations/data_context/data_context/base_data_context.py
Comment (cdkini) [inline great_expectations/data_context/data_context/base_data_context.py:189]: Moved from a static method to a stand-alone helper

File: great_expectations/data_context/data_context/data_context.py
Comment (cdkini) [inline great_expectations/data_context/data_context/data_context.py]: Took any class methods that were really specific to `DataContext` and made them free-standing helpers - these are usually methods that combine file and Cloud behavior (but don't fit in `SerializableDataContext`, `CloudDataContext`, or `FileDataContext`)

File: great_expectations/data_context/data_context/explorer_data_context.py
Comment (cdkini) [inline great_expectations/data_context/data_context/explorer_data_context.py]: Will come back to this 😕

File: great_expectations/data_context/data_context/file_data_context.py
Comment (cdkini) [inline great_expectations/data_context/data_context/file_data_context.py:42]: If not present, we should have legacy `DataContext` behavior where we parse it from the filesystem

File: azure-pipelines-dev.yml
Comment (cdkini) [inline azure-pipelines-dev.yml]: Will resolve all of these after the first round of review - should mainly be just `grep` and `sed` to convert old types to `AbstractDataContext` and its subclasses

File: great_expectations/checkpoint/actions.py
Comment (cdkini) [inline great_expectations/checkpoint/actions.py]: We'll want to clean this up eventually but Cloud leaks into everything so this is one way to encapsulate it a bit

File: great_expectations/cli/cli.py
Comment (cdkini) [inline great_expectations/cli/cli.py:54]: Type hints will be updated so please ignore those for now

File: great_expectations/data_context/data_context/abstract_data_context.py
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py:582]: Ported over from `DataContext`

File: great_expectations/data_context/data_context/explorer_data_context.py
Comment (cdkini) [reply great_expectations/data_context/data_context/explorer_data_context.py]: Think we can just delete this but maybe I'm being too rash

File: great_expectations/data_context/data_context/serializable_data_context.py
Comment (cdkini) [inline great_expectations/data_context/data_context/serializable_data_context.py]: Required to fulfill ABC but we never actually instantiate a `SerializableDataContext` by itself - thoughts?

File: great_expectations/data_context/data_context/serializable_data_context.py
Comment (cdkini) [inline great_expectations/data_context/data_context/serializable_data_context.py:35]: TODO: docstring (and bring back any old docstrings I may have deleted)

File: great_expectations/data_context/data_context/serializable_data_context.py
Comment (cdkini) [inline great_expectations/data_context/data_context/serializable_data_context.py]: Needs deletion

File: great_expectations/data_context/data_context/serializable_data_context.py
Comment (cdkini) [inline great_expectations/data_context/data_context/serializable_data_context.py:84]: All these methods are coming over from `DataContext`

File: great_expectations/checkpoint/actions.py
Comment (Kilo59) [reply great_expectations/checkpoint/actions.py]: Why not just keep the existing `self.data_context.cloud_mode` until we can improve the abstraction?
This change seems lateral.

File: great_expectations/data_context/data_context/file_data_context.py
Comment (Kilo59) [inline great_expectations/data_context/data_context/file_data_context.py]:
```suggestion
        path_to_yml = pathlib.Path(context_root_directory, cls.GX_YML)
```
Let's use `pathlib` here, and remove the `os` import.

File: great_expectations/rule_based_profiler/rule_based_profiler.py
Comment (Kilo59) [inline great_expectations/rule_based_profiler/rule_based_profiler.py]: Aren't these the same check?

File: great_expectations/validation_operators/validation_operators.py
Comment (Kilo59) [inline great_expectations/validation_operators/validation_operators.py]: Same comment as above. Why are we making these `self.data_context.cloud_mode -> if self._cloud_mode` changes?

File: tests/data_context/test_data_context.py
Comment (Kilo59) [inline tests/data_context/test_data_context.py]: Probably already on your radar but, don't forget about this.

File: tests/data_context/test_data_context.py
Comment (Kilo59) [inline tests/data_context/test_data_context.py]: Can you annotate the type of this fixture?
Is it `pathlib.Path` or a `str`?

File: great_expectations/data_context/data_context/serializable_data_context.py
Comment (Kilo59) [inline great_expectations/data_context/data_context/serializable_data_context.py:173]: **non-blocking/followup**
Since this is a net new module can we use `pathlib` and remove `os`?
If compatibility is a concern we can just call `str(my_path_obj)` on the path object when passing it.

Here's a handy table that maps `os.path` methods to `pathlib` methods.
https://docs.python.org/3/library/pathlib.html#correspondence-to-tools-in-the-os-module

File: great_expectations/data_context/data_context/serializable_data_context.py
Comment (Kilo59) [inline great_expectations/data_context/data_context/serializable_data_context.py:115]: Can we fix these issues on `AnonymizedUsageStatisticsConfig`?
Assuming it's a small change.

File: great_expectations/checkpoint/actions.py
Comment (cdkini) [reply great_expectations/checkpoint/actions.py]: not all contexts have cloud_mode

File: tests/data_context/test_data_context.py
Comment (cdkini) [reply tests/data_context/test_data_context.py]: yup going to delete!

File: tests/data_context/test_data_context.py
Comment (cdkini) [reply tests/data_context/test_data_context.py]: it's a `pytest` built-in: `_pytest.fixture.TemporaryPathFactory` I believe

File: great_expectations/data_context/data_context/file_data_context.py
Comment (cdkini) [reply great_expectations/data_context/data_context/file_data_context.py]: done!

File: great_expectations/validation_operators/validation_operators.py
Comment (cdkini) [reply great_expectations/validation_operators/validation_operators.py]: same note as elsewhere - `cloud_mode` is no longer a property so we're doing an `isinstance` check

File: great_expectations/data_context/data_context/serializable_data_context.py
Comment (cdkini) [reply great_expectations/data_context/data_context/serializable_data_context.py:173]: I'd be happy to but would you mind if I did this `os` replacement work in a subsequent PR? Can add to the context epic I'm working in but I'd like to focus on getting this change in before it gets stale.

Will defer to you here.

File: great_expectations/data_context/data_context/serializable_data_context.py
Comment (cdkini) [reply great_expectations/data_context/data_context/serializable_data_context.py:115]: We've got a ticket in the context epic to resolve `mypy` issues - mind if we tackle that there?

File: great_expectations/data_context/data_context/serializable_data_context.py
Comment (Kilo59) [reply great_expectations/data_context/data_context/serializable_data_context.py:173]: Sounds good to me.

File: great_expectations/checkpoint/actions.py
Comment (Kilo59) [inline great_expectations/checkpoint/actions.py]: It's still not clear to me why it's not better to have the base/abstract `DataContext` class have a `data_context.cloud_mode` attribute that defaults to false.
I don't think the extra code churn here actually makes the abstraction any better. But maybe I just don't understand it.

It should go without saying that we should work to abstract away these conditional `if cloud_mode` checks in a subsequent effort.

File: great_expectations/data_context/data_context/serializable_data_context.py
Comment (Kilo59) [reply great_expectations/data_context/data_context/serializable_data_context.py:115]: That works for me.

File: great_expectations/checkpoint/actions.py
Comment (billdirks) [reply great_expectations/checkpoint/actions.py]: One goal of this effort is to remove the notion of `cloud` from the data contexts except for the `CloudDataContext`. I think that is what is motivating this. Are we removing this too early, since now we must do checks like this in 3 files? maybe. We should remove these checks in general and not need to know if we are working with a cloud context at all.

File: great_expectations/checkpoint/checkpoint.py
Comment (billdirks) [inline great_expectations/checkpoint/checkpoint.py:498]: I'm guessing we never have a true `AbstractDataContext` here but a subclass. What you are doing is fine. You could use a Generic here. That is at the top of the file do: ` T = TypeVar('T', bound=AbstractDataContext)`. Then in the class declaration you'd do `class BaseCheckpoint(ConfigPeer, Generic[T]):` and replace all instances of `AbstractDataContext` by `T`. This means if the class was initialized with a `FileDataContext`, the type checker would then know that the getter will return a `FileDataContext` and not an abstract one.

This comment doesn't block this PR. I'm trying to highlight a reason you might want to use generics here.

File: great_expectations/data_context/data_context/abstract_data_context.py
Comment (billdirks) [reply great_expectations/data_context/data_context/abstract_data_context.py:166]: `ConfigPeer` is an `ABC` so we do need that here? Previously did all previous `*DataContext`s inherit from `ConfigPeer`? I'm just wondering the ramifications of adding `ConfigPeer` as a parent class here.

**Review by billdirks — COMMENTED** [review]
Thanks for doing this. I've only looked at a quarter of the files and will do more tomorrow.

File: great_expectations/checkpoint/actions.py
Comment (cdkini) [reply great_expectations/checkpoint/actions.py]: These checks definitely need to go but as it stands, there is quite a bit of Cloud leakage into other parts of the codebase. Since this property is simply encapsulating that leakage here, I'd vote to keep this for now and then work to decouple Cloud in a subsequent effort.

Just to clarify, we're not adding a new behavior here - instead of checking for `cloud_mode`, we're evaluating whether or not we're in a Cloud backed environment through the context type (different ways to achieve the same goal).

Happy to brainstorm a better way to do this.

File: great_expectations/checkpoint/checkpoint.py
Comment (cdkini) [reply great_expectations/checkpoint/checkpoint.py:498]: nope this is a great point and something I think should go in as part of the implementation - I've paired with @Kilo59 just to clarify some points and have made the necessary changes.

Let me know if you agree!

File: great_expectations/checkpoint/actions.py
Comment (Kilo59) [reply great_expectations/checkpoint/actions.py]: @cdkini I understand that we aren't adding new behavior here. I also accept that this sort of check is still needed.

What I'm saying is that I don't think only having the `cloud_mode` attribute on the `CloudDataContext` really improves things, since you still have to check for it on every `DataContext` anyways. Except now it's harder because we are wary of `AttributeErrors`.

Comment (cdkini) [reply great_expectations/checkpoint/actions.py]: oh no we removed `cloud_mode` as an attr now - hence the change to `isinstance` as a check

Comment (Kilo59) [reply great_expectations/checkpoint/actions.py]: >oh no we removed cloud_mode as an attr now - hence the change to isinstance as a check

I understand that `cloud_mode` is only an attr on  `CloudDataContext`.

File: great_expectations/validation_operators/validation_operators.py
Comment (cdkini) [inline great_expectations/validation_operators/validation_operators.py]: Change name to `is_cloud_context` or something a bit different to better convey what is going on with this temp property?

File: great_expectations/data_context/data_context/serializable_data_context.py
Comment (billdirks) [reply great_expectations/data_context/data_context/serializable_data_context.py]: Are you getting a linting check since it expects this class to be concrete? I wonder if you can tell whichever check is raising the issue that this class is still abstract?

Otherwise, you could do `raise NotImplementedError` instead of passing so no one will calls these on accident.

File: great_expectations/data_context/data_context/serializable_data_context.py
Comment (billdirks) [inline great_expectations/data_context/data_context/serializable_data_context.py]: Is there any helpful info you can provide about the DataContext here? I'm fine with keeping it, but I'm wondering how useful this debug message is.

**Review by billdirks — COMMENTED** [review]
Thanks for doing this refactor @cdkini! I have one small comment about understanding the linting error that can be resolved by removing the `pass` implementations by telling the linter it's ok that the methods aren't there or by implementing them with a `raise NotImplementedError`.

Otherwise I'm fine with this change and you can @Kilo59 can work through the discussion you are having.

Thanks!

File: great_expectations/data_context/data_context/serializable_data_context.py
Comment (cdkini) [reply great_expectations/data_context/data_context/serializable_data_context.py]: done!

**Review by Kilo59 — APPROVED** [review]

**Pull request merged**
